### ISSUE TITLE
Test failures with Arrow 18

### ISSUE THREAD

**Issue by tobim** [OSGeo/gdal#11163]
### What is the bug?

When trying to build and test gdal 3.9.3 with Apache Arrow 18.0.0 two tests fail.
I presume this is caused by a behavior change in the metadata extraction for empty
or faulty row groups. When building with Arrow 17.0.0 these tests pass.

Full error log:
```
=================================== FAILURES ===================================
______________________ test_ogr_parquet_write_edge_cases _______________________

    def test_ogr_parquet_write_edge_cases():
    
        outfilename = "/vsimem/out.parquet"
    
        # No layer
        ds = gdal.GetDriverByName("Parquet").Create(outfilename, 0, 0, 0, gdal.GDT_Unknown)
        assert ds is not None
        assert ds.GetLayerCount() == 0
        assert ds.GetLayer(0) is None
        assert ds.TestCapability(ogr.ODsCCreateLayer) == 1
        assert ds.TestCapability(ogr.ODsCAddFieldDomain) == 0
        domain = ogr.CreateCodedFieldDomain(
            "name", "desc", ogr.OFTInteger, ogr.OFSTNone, {1: "one", "2": None}
        )
        assert ds.AddFieldDomain(domain) == False
        assert ds.GetFieldDomainNames() is None
        assert ds.GetFieldDomain("foo") is None
        ds = None
        gdal.Unlink(outfilename)
    
        # No field, no record
        ds = gdal.GetDriverByName("Parquet").Create(outfilename, 0, 0, 0, gdal.GDT_Unknown)
        assert ds is not None
        srs = osr.SpatialReference()
        srs.ImportFromEPSG(4326)
        with pytest.raises(Exception):
            ds.CreateLayer("out", srs=srs, geom_type=ogr.wkbPointM)
        with pytest.raises(Exception):
            ds.CreateLayer(
                "out", srs=srs, geom_type=ogr.wkbPoint, options=["COMPRESSION=invalid"]
            )
        lyr = ds.CreateLayer("out", srs=srs, geom_type=ogr.wkbPoint)
        assert lyr is not None
        assert lyr.GetDataset().GetDescription() == ds.GetDescription()
        assert ds.GetLayerCount() == 1
        assert ds.GetLayer(0) is not None
        assert ds.TestCapability(ogr.ODsCCreateLayer) == 0
        assert ds.TestCapability(ogr.ODsCAddFieldDomain) == 1
        # Test creating a second layer
        with pytest.raises(Exception):
            ds.CreateLayer("out2", srs=srs, geom_type=ogr.wkbPoint)
        ds = None
        ds = gdal.OpenEx(outfilename)
        assert ds is not None
        lyr = ds.GetLayer(0)
>       assert lyr.GetNextFeature() is None

/build/source/build/autotest/ogr/ogr_parquet.py:688: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <osgeo.ogr.Layer; proxy of <Swig Object of type 'OGRLayerShadow *' at 0x7ffee0aa96e0> >
args = ()

    def GetNextFeature(self, *args):
        r"""
        GetNextFeature(Layer self) -> Feature
    
        Fetch the next available feature from this layer.
    
        For more details: :cpp:func:`OGR_L_GetNextFeature`
    
        Returns
        --------
        Feature:
            A feature or None if no more features are available.
    
        """
>       return _ogr.Layer_GetNextFeature(self, *args)
E       RuntimeError: ReadNext() failed: The file only has 0 row groups, requested metadata for row group: -1
E       /build/source/cpp/src/parquet/arrow/reader.cc:109  LoadBatch(batch_size)
E       /build/source/cpp/src/parquet/arrow/reader.cc:1037  ::arrow::internal::OptionalParallelFor( reader_properties_.use_threads(), static_cast<int>(readers.size()), [&](int i) { return readers[i]->NextBatch(batch_size, &columns[i]); })
E       /build/source/cpp/src/arrow/util/iterator.h:537  parent_.Next()

/nix/store/dicp2s6g7s2v7anv2wvxgzcmkwm3sdfi-gdal-3.9.3/lib/python3.12/site-packages/osgeo/ogr.py:1679: RuntimeError
----------------------------- Captured stderr call -----------------------------
Warning 6: 'invalid' is an unexpected value for COMPRESSION layer creation option of type string-select.
___________________ test_ogr_parquet_arrow_stream_empty_file ___________________

    def test_ogr_parquet_arrow_stream_empty_file():
    
        ds = ogr.GetDriverByName("Parquet").CreateDataSource("/vsimem/test.parquet")
        ds.CreateLayer("test", geom_type=ogr.wkbPoint)
        ds = None
        ds = ogr.Open("/vsimem/test.parquet")
        lyr = ds.GetLayer(0)
        assert lyr.TestCapability(ogr.OLCFastGetArrowStream) == 1
        stream = lyr.GetArrowStream()
>       assert stream.GetNextRecordBatch() is None

/build/source/build/autotest/ogr/ogr_parquet.py:1891: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <osgeo.ogr.ArrowArrayStream; proxy of <Swig Object of type 'ArrowArrayStream *' at 0x7ffed9132eb0> >
args = ()

    def GetNextRecordBatch(self, *args):
        r"""GetNextRecordBatch(ArrowArrayStream self, char ** options=None) -> ArrowArray"""
>       return _ogr.ArrowArrayStream_GetNextRecordBatch(self, *args)
E       RuntimeError: ReadNext() failed: The file only has 0 row groups, requested metadata for row group: -1
E       /build/source/cpp/src/parquet/arrow/reader.cc:109  LoadBatch(batch_size)
E       /build/source/cpp/src/parquet/arrow/reader.cc:1037  ::arrow::internal::OptionalParallelFor( reader_properties_.use_threads(), static_cast<int>(readers.size()), [&](int i) { return readers[i]->NextBatch(batch_size, &columns[i]); })
E       /build/source/cpp/src/arrow/util/iterator.h:537  parent_.Next()

/nix/store/dicp2s6g7s2v7anv2wvxgzcmkwm3sdfi-gdal-3.9.3/lib/python3.12/site-packages/osgeo/ogr.py:1424: RuntimeError
=========================== short test summary info ============================
FAILED ogr/ogr_parquet.py::test_ogr_parquet_write_edge_cases - RuntimeError: ReadNext() failed: The file only has 0 row groups, requested metadata for row group: -1
/build/source/cpp/src/parquet/arrow/reader.cc:109  LoadBatch(batch_size)
/build/source/cpp/src/parquet/arrow/reader.cc:1037  ::arrow::internal::OptionalParallelFor( reader_properties_.use_threads(), static_cast<int>(readers.size()), [&](int i) { return readers[i]->NextBatch(batch_size, &columns[i]); })
/build/source/cpp/src/arrow/util/iterator.h:537  parent_.Next()
FAILED ogr/ogr_parquet.py::test_ogr_parquet_arrow_stream_empty_file - RuntimeError: ReadNext() failed: The file only has 0 row groups, requested metadata for row group: -1
/build/source/cpp/src/parquet/arrow/reader.cc:109  LoadBatch(batch_size)
/build/source/cpp/src/parquet/arrow/reader.cc:1037  ::arrow::internal::OptionalParallelFor( reader_properties_.use_threads(), static_cast<int>(readers.size()), [&](int i) { return readers[i]->NextBatch(batch_size, &columns[i]); })
/build/source/cpp/src/arrow/util/iterator.h:537  parent_.Next()
= 2 failed, 12876 passed, 1398 skipped, 7 deselected, 6 xfailed in 583.24s (0:09:43) =
100 - done.
```

### Steps to reproduce the issue

If you have Nix available:
`nix --extra-experimental-features "nix-command flakes" -L build github:cpcloud/nixpkgs/arrow-18#gdal`

### Versions and provenance

NixOS 24.05.
GDAL 3.9.3.

**Comment by rouault** [comment]
yes, fixed in 3.10.0rc1 per https://github.com/OSGeo/gdal/commit/9a8c5c031404bbc81445291bad128bc13766cafa

The minimu bugfix is actually the changes in ogrparquetlayer.cpp: https://github.com/OSGeo/gdal/commit/9a8c5c031404bbc81445291bad128bc13766cafa#diff-c0ee9dfe06d693d738ec8ae37931f494d3a7812b7d19730ab5aeb7928ba1fac0

**Comment by rouault** [comment]
See also this email "GDAL 3.9 updates for SWIG 4.3 and libarrow 18": https://lists.osgeo.org/pipermail/gdal-dev/2024-October/059713.html
